### PR TITLE
FeeBurner contract implementation

### PR DESCRIPTION
**Pull request by pik694**
Implemented fee-burner's contract with its tests in python.

### CODE REVIEW

File: contracts/FeeBurner.sol [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+    {
+            // Checks whether such a token hasn't been supported yet
+            require(exchangeRates[_token].blockNo == 0);
+
+            ExchangeRate memory exchangeRate = ExchangeRate(block.number, Rate(_nominator, _denominator));
+
+            previousExchangeRates[_token] = exchangeRate;
+            exchangeRates[_token] = exchangeRate;
+
+            //TODO: Should I emit an event ? Maybe sth like AddedSupport(token)
```
Comment (pdobacz) [inline contracts/FeeBurner.sol]: yes and no. I think `ExchangeRateChanged` on both occasions is fine, the effect is the same, so apps tracking the exchange rates will have an easier job

File: contracts/FeeBurner.sol [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+     * @param _nominator nominator of the rate at which the user wants to make the exchange
+     * @param _denominator denominator of the rate at which the user wants to make the exchange
+     * @param _omg_amount amount of OMGs to be exchanged and burnt
+     * @param _token_amount amount of ERC20 tokens to be exchanged
+     *
+     * @notice Rate proposed by the user must equal either to exchangeRate or to previousExchangeRate.
+     * @notice Rate cannot equal to the previousExchangeRate when the maturity period of the new rate has passed.
+     * @notice Sender may offer more OMGs than demanded by the exchange rate, the exchange/transaction is then valid.
+     */
+    function exchange(address _token, uint _nominator, uint _denominator, uint _omg_amount, uint _token_amount)
```
Comment (pdobacz) [inline contracts/FeeBurner.sol]: I'd name the arguments _rate_nominator and _rate_denominator for a little more readability.

File: contracts/FeeBurner.sol [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+            msg.sender.transfer(_token_amount);
+        }
+        else {
+            ERC20 token = ERC20(_token);
+            token.transfer(msg.sender, _token_amount);
+        }
+
+    }
+
+    //TODO: should I leave this convenience method ?
```
Comment (pdobacz) [inline contracts/FeeBurner.sol]: if in doubt, leave it out

File: contracts/FeeBurner.sol [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+    }
+
+    /*
+     * Public view functions
+     */
+
+    /**
+     * @dev Returns flat pending exchange rate of the given token. 
+     * 
+     * @param _token address of an ERC20 token, or 0 in case of Ethereum
```
Comment (pdobacz) [inline contracts/FeeBurner.sol]: the return value is most interesting here

File: tests/test_fee_burner.py [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+    compiled_fee_burner = compile_source_file('./contracts/FeeBurner.sol')
+    contract_interface = compiled_fee_burner['./contracts/FeeBurner.sol:FeeBurner']
+
+    address = deploy_contract(w3, contract_interface, omg_token.address)
+    fee_burner_contract = w3.eth.contract(
+        address=address,
+        abi=contract_interface['abi']
+    )
+
+    fee_burner_contract.transact().addSupportFor(ZERO_ADDRESS, 1, 1)
```
Comment (pdobacz) [inline tests/test_fee_burner.py]: are you sure you want to do that on every fee burner fixture instance?

File: tests/test_fee_burner.py [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+    fee_burner.functions.addSupportFor(other_token.address, 1, 123).transact()
+
+    # then: this token is supported
+    assert fee_burner.functions.getExchangeRate(other_token.address).call()[-2:] == [1, 123]
+
+
+def test_add_support_for_a_token_by_a_non_operator(fee_burner, non_operator, other_token):
+    # when: a non operator tries to add support for a token
+    # then: an error occurs
+    with pytest.raises(ValueError):  # TODO: now ValueError is thrown instead of TransactionFailed
```
Comment (pdobacz) [inline tests/test_fee_burner.py]: this is a little bit suspicious, but maybe that's what w3 throws on error in contract execution. Expecting a specific `TransactionFailed` seemed safer...

File: tests/test_fee_burner.py [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+    mine_blocks(provider, fee_burner.functions.NEW_RATE_MATURITY_MARGIN().call() + 1)
+    block_no = w3.eth.blockNumber
+
+    # then: operator can once more change the exchange rate
+    fee_burner.functions.setExchangeRate(ZERO_ADDRESS, 22, 33).transact()
+    assert fee_burner.functions.getExchangeRate(ZERO_ADDRESS).call() == [block_no + 1, 22, 33]
+    assert fee_burner.functions.getPreviousExchangeRate(ZERO_ADDRESS).call() == changed_rate
+
+
+def test_during_maturity_period_both_rates_should_be_valid(provider, non_operator, fee_burner, other_token, omg_token):
```
Comment (pdobacz) [inline tests/test_fee_burner.py]: you could add some tests where you test various combinations of rate_nominator, rate_denominator, omg_amount, other_amount - to make sure that the trades done always respect the exchange rate, i.e. when buing a token valued higher/lower/equal to omg

File: tests/test_fee_burner.py [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+    # then: error
+    with pytest.raises(ValueError):
+        fee_burner.functions.exchange(other_token.address, 2, 1, 2, 1).transact({'from': non_operator})
+
+    # when: the user sends an exchange demand OMG for other token at valid rate, but amounts aren't at this rate
+    # then: error
+    with pytest.raises(ValueError):
+        fee_burner.functions.exchange(other_token.address, 1, 1, 1, 2).transact({'from': non_operator})
+
+    # TODO: probably I needn't check this
```
Comment (pdobacz) [inline tests/test_fee_burner.py]: I think I wouldn't. I would only if you did a similar maneouver as in HonteStaking tests, where the `transacts` were wrapped and the wrappers checked all the post conditions after a transaction. HonteStaking had this only for successful transactions, but you can also have a failure-expecting wrapper, that works the similar way. Such wrapper would then check post-conditions like these.

File: tests/test_fee_burner.py (hunk 6 above)
Comment (pik694) [reply tests/test_fee_burner.py]: `TransactionFailed` is purely `eth-tester`'s exception and I do not use this package. This is what `web3` throws. A workaround could be to use the proposed wrapper (d060b75dbbb980f1b0b2254a38339c1bf5c14a4f) of the contract and in its implementation I would catch `ValueError` and throw `TransactionFailed` exception.

Comment (pdobacz) [reply tests/test_fee_burner.py]: the point is that `ValueError` is too generic, and in tests you'd like to be sure that you're action is throwing a specific exception that indicates that the transaction failed. I think that in `ValueError` there's a msg field that elaborates on the reason of failure, that maybe you could check. You'll then have to write your own `with` syntax to wrap around `with py.test raises`.

ah, I remember @4gn3s did it once for a former project, there it is:

```
class RaisesTransactionFailed:
    # see https://github.com/pipermerriam/web3.py/issues/75
    def __enter__(self):
        return self

    def __exit__(self, exception_type, exception_value, traceback):
        with pytest.raises(TransactionFailed):
            if exception_type == ValueError and exception_value[0]['data']['type'] == u'TransactionFailed':
                raise TransactionFailed
        return True
```

then
```
       with RaisesTransactionFailed():
            contract.transact()...
```

**Pull request merged**